odl/operator/fspace_ops.py

```diff
@@ -0,0 +1,412 @@
+    array([ 0.5,  1. ,  1. ,  1. ,  1. ,  1. ,  1. ,  1. ,  1. , -4. ])
+    >>> df1 = discrete_part_deriv(np.sin(f/10*np.pi), edge_order=1)
+    >>> df2 = discrete_part_deriv(np.sin(f/10*np.pi), edge_order=2)
+    >>> np.array_equal(df1[1:-1], df2[1:-1])
+    True
+    >>> df1[0] == df2[0]
+    False
+    >>> df1[-1] == df2[-1]
+    False
+    >>> n = 5
+    >>> f = np.arange(n, dtype=float)
+    >>> f = f * f.reshape((n,1))
+    >>> discrete_part_deriv(f, 0)
+    array([[-0.,  1.,  2.,  3.,  4.],
+           [ 0.,  1.,  2.,  3.,  4.],
+           [ 0.,  1.,  2.,  3.,  4.],
+           [ 0.,  1.,  2.,  3.,  4.],
+           [ 0.,  1.,  2.,  3.,  4.]])
+    >>> discrete_part_deriv(f, 1)
+    array([[-0.,  0.,  0.,  0.,  0.],
+           [ 1.,  1.,  1.,  1.,  1.],
+           [ 2.,  2.,  2.,  2.,  2.],
+           [ 3.,  3.,  3.,  3.,  3.],
+           [ 4.,  4.,  4.,  4.,  4.]])
+    >>> try:
+    ...    discrete_part_deriv(f, 2)
+    ... except IndexError as e:
+    ...    print(e)
+    Axis paramater (2) exceeds number of dimensions (2).
+    """
+
+    f_data = np.asanyarray(f)
+    ndim = f_data.ndim
+
+    # create slice objects --- initially all are [:, :, ..., :]
+    # noinspection PyTypeChecker
+    slice1 = [slice(None)] * ndim
+    # noinspection PyTypeChecker
+    slice2 = [slice(None)] * ndim
+    # noinspection PyTypeChecker
+    slice3 = [slice(None)] * ndim
+    # noinspection PyTypeChecker
+    slice4 = [slice(None)] * ndim
+
+    try:
+        if f_data.shape[axis] < 2:
+            raise ValueError("Shape of array too small to calculate a "
+                             "numerical gradient, at least two elements are "
+                             "required.")
+    except IndexError:
+        raise IndexError("Axis paramater ({0}) exceeds number of dimensions "
+                         "({1}).".format(axis, ndim))
+
+    out = np.empty_like(f_data)
+
+    # Numerical differentiation: 2nd order interior
+    slice1[axis] = slice(1, -1)
+    slice2[axis] = slice(2, None)
+    # noinspection PyTypeChecker
+    slice3[axis] = slice(None, -2)
+    # 1D equivalent -- out[1:-1] = (y[2:] - y[:-2])/2.0
+    out[slice1] = (f_data[slice2] - f_data[slice3]) / 2.0
+
+    # central differences
+    if zero_padding:
+        # Assume zeros for indices outside the volume
+
+        # 1D equivalent -- out[0] = (y[1] - 0)/2.0
+        slice1[axis] = 0
+        slice2[axis] = 1
+        out[slice1] = f_data[slice2] / 2.0
+
+        # 1D equivalent -- out[-1] = (0 - y[-2])/2.0
+        slice1[axis] = -1
+        slice3[axis] = -2
+        out[slice1] = - f_data[slice3] / 2.0
+
+    # one-side differences
+    else:
+        # Numerical differentiation: 1st order edges
+        if f_data.shape[axis] == 2 or edge_order == 1:
+
+            slice1[axis] = 0
+            slice2[axis] = 1
+            slice3[axis] = 0
+            # 1D equivalent -- out[0] = (y[1] - y[0])
+            out[slice1] = (f_data[slice2] - f_data[slice3])
```

Comment:
Yeah, its `np.subtract`. Using the `out` parameter you dont need to make a copy. Both faster and more memory efficient.

Your current code is:

```python
tmp = f_data[slice2] - f_data[slice3]
out[slice1] = tmp
```